### PR TITLE
Hardware factory

### PR DESCRIPTION
**Pull request by kress95**
Mainly adds factories for Hardware, but also fixes some other problems and adds "detach all components from a motherboard" controller function.

---
This change is [Reviewable](https://reviewable.io/reviews/hackerexperience/helix/5)

### CODE REVIEW

**Comment by kress95** [comment]
There are some remaining tests to add, but I already started to add them on another branch.
I believe it's better to just ensure that this is fixed and ship it asap, the sooner this ships, the sooner I send new revisions with the things we need.

**Comment by kress95** [comment]
- previously reviewed by me before asking this additional review
- tested many times
- strictly analyzed by both credo and dialyzer
- fully rebased with master
- minor tests style fixes
- hardware factory refactor, it's easier to understand and extend now

**Comment by mememori** [comment]
Reviewed 2 of 9 files at r1, 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[apps/helix_hardware/test/support/factory.ex, line 38 at r2](https://reviewable.io:443/reviews/hackerexperience/helix/5#-KgHbmZoG-w2m2dGjj_w:-KgHbmZoG-w2m2dGjj_x:b200qt1) ([raw file](https://github.com/hackerexperience/helix/blob/24cbbdddf010531c98b7e8ea9eb939b9b0d93f2d/apps/helix_hardware/test/support/factory.ex#L38)):*
> ```Elixir
>       Enum.map(specs, fn {id, spec} ->
>         %MotherboardSlot{
>           slot_id: PK.pk_for(MotherboardSlot),
> ```

... Wasn't this being autogenerated by ecto ?

---

*[apps/helix_hardware/test/support/factory.ex, line 112 at r2](https://reviewable.io:443/reviews/hackerexperience/helix/5#-KgHc3ydSLoZdDpPp8M5:-KgHc3ydSLoZdDpPp8M6:bc65zsh) ([raw file](https://github.com/hackerexperience/helix/blob/24cbbdddf010531c98b7e8ea9eb939b9b0d93f2d/apps/helix_hardware/test/support/factory.ex#L112)):*
> ```Elixir
>   defp prepare(:component, params) do
>     spec = Keyword.get(params, :component_spec,
>       prepare(:component_spec, params))
> ```

why did you indent like this ?

---

*[apps/helix_hardware/test/support/factory.ex, line 117 at r2](https://reviewable.io:443/reviews/hackerexperience/helix/5#-KgHc8YiOSMgND-PAZMU:-KgHc8YiOSMgND-PAZMV:bhpeh0f) ([raw file](https://github.com/hackerexperience/helix/blob/24cbbdddf010531c98b7e8ea9eb939b9b0d93f2d/apps/helix_hardware/test/support/factory.ex#L117)):*
> ```Elixir
>       ComponentType.type_implementations()
>       |> Map.fetch!(spec.component_type)
>       |> PK.pk_for()
> ```

nice hack

---

*[apps/helix_hardware/test/support/factory.ex, line 140 at r2](https://reviewable.io:443/reviews/hackerexperience/helix/5#-KgHcjx40slGUsWu26ra:-KgHcjx40slGUsWu26rb:b-9dl63r) ([raw file](https://github.com/hackerexperience/helix/blob/24cbbdddf010531c98b7e8ea9eb939b9b0d93f2d/apps/helix_hardware/test/support/factory.ex#L140)):*
> ```Elixir
>     ComponentType.type_implementations()
>     |> Map.keys()
>     |> Enum.reject(&(&1 == "mobo"))
> ```

why are you rejecting motherboard components ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/helix/5)*

**Comment by kress95** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[apps/helix_hardware/test/support/factory.ex, line 38 at r2](https://reviewable.io:443/reviews/hackerexperience/helix/5#-KgHbmZoG-w2m2dGjj_w:-KgJjTnz0oVaHzwCdxJy:b-pubwzi) ([raw file](https://github.com/hackerexperience/helix/blob/24cbbdddf010531c98b7e8ea9eb939b9b0d93f2d/apps/helix_hardware/test/support/factory.ex#L38)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

... Wasn't this being autogenerated by ecto ?
</blockquote></details>

I'll check it, it might passed unperceived.

---

*[apps/helix_hardware/test/support/factory.ex, line 112 at r2](https://reviewable.io:443/reviews/hackerexperience/helix/5#-KgHc3ydSLoZdDpPp8M5:-KgJjcEr7_qInZj3Rduu:b-5wwxao) ([raw file](https://github.com/hackerexperience/helix/blob/24cbbdddf010531c98b7e8ea9eb939b9b0d93f2d/apps/helix_hardware/test/support/factory.ex#L112)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

why did you indent like this ?
</blockquote></details>

pls suggest  better way to indent it, I'll follow the style everywhere

---

*[apps/helix_hardware/test/support/factory.ex, line 117 at r2](https://reviewable.io:443/reviews/hackerexperience/helix/5#-KgHc8YiOSMgND-PAZMU:-KgJjlxiHF0zBcoFELR-:bm2m8rl) ([raw file](https://github.com/hackerexperience/helix/blob/24cbbdddf010531c98b7e8ea9eb939b9b0d93f2d/apps/helix_hardware/test/support/factory.ex#L117)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

nice hack
</blockquote></details>

thx :joy: 

---

*[apps/helix_hardware/test/support/factory.ex, line 140 at r2](https://reviewable.io:443/reviews/hackerexperience/helix/5#-KgHcjx40slGUsWu26ra:-KgJk1XXs8BMWtiY3HPi:bt74pmq) ([raw file](https://github.com/hackerexperience/helix/blob/24cbbdddf010531c98b7e8ea9eb939b9b0d93f2d/apps/helix_hardware/test/support/factory.ex#L140)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

why are you rejecting motherboard components ?
</blockquote></details>

This was done defensively as we usually get components to test slot linking, but it's wrong to assume that, I'll remove this constraint

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/helix/5)*

**Comment by kress95** [comment]
I just remember something, on some previous revision (maybe this one) you criticized `prepare/1 | prepare/2` as they looked a lot like native ExMachina functions.

I don't know if this still's a thing as you didn't even mention it now, but I'll happily change `prepare` name to something more obviously "not ExMachina' if that's the case.

**Comment by mememori** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/helix/5#-:-KgKXOQo-A0r15GZSHyj:bnfp4nl)*

File: apps/hardware/test/controller/motherboard_test.exs [hunk 1]
```diff
@@ -34,12 +34,9 @@ defmodule Helix.Hardware.Controller.MotherboardTest do
...
     MotherboardController.unlink_components_from_motherboard(mobo)
 
-    non_empty_slots =
-      mobo
-      |> MotherboardController.get_slots()
-      |> Enum.reject(&is_nil(&1.link_component_id))
+    unused_slot? = &is_nil(&1.link_component_id)
 
-    assert [] == non_empty_slots
+    assert Enum.all?(MotherboardController.get_slots(mobo), unused_slot?)
```
Comment (kress95) [inline apps/hardware/test/controller/motherboard_test.exs]: This one escaped D:

**Pull request merged**